src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+                    if (NeighbourUtils.up(cellIndex).equals(neighbourIndex)) {
+                        diff = -Math.min(oldCell.smokeIndicator(), 100 - oldState.getCell(neighbourIndex).smokeIndicator());
+                        counter.v += diff;
+                    } else if (NeighbourUtils.down(cellIndex).equals(neighbourIndex)) {
+                        diff = Math.min(oldState.getCell(neighbourIndex).smokeIndicator(),
+                                100 - oldCell.smokeIndicator());
+                        counter.v += diff;
+                    } else if (isHorizontal(cellIndex, neighbourIndex)) {
+                        int numberOfNeighboursMine = N(oldState, cellIndex).size() + 1;
+                        int numberOfNeighboursHim = N(oldState, neighbourIndex).size() + 1;
+                        if (cellAboveCanNotTakeSmoke(oldState, cellIndex)) {
+                            diff -= Math.min((100 - oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim,
+                                    (counter.v) / numberOfNeighboursMine);
+                        }
+                        if (cellAboveCanNotTakeSmoke(oldState, neighbourIndex)) {
+                            diff += Math.min((100 - counter.v) / numberOfNeighboursMine,
+                                    (oldState.getCell(neighbourIndex).smokeIndicator()) / numberOfNeighboursHim);
+                        }
+                    }
+                    return diff;
+                }).mapToDouble(Double::doubleValue).sum();
+    }
+
+    private boolean cellAboveCanNotTakeSmoke(State state, Vector3i index) {
+        Vector3i cellAbove = NeighbourUtils.up(index);
+        return (state.hasCell(cellAbove) && state.getCell(cellAbove).smokeIndicator() >= 50) ||
+                (state.hasCell(cellAbove) && !state.getCell(cellAbove).isFluid()) ||
+                !state.hasCell(cellAbove);
+    }
+
+    private List<Cell> N(State state, Vector3i cellIndex) {
+        List<Cell> res = new ArrayList<>();
+        NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.X)
+                .filter(state::hasCell)
+                .filter(index -> state.getCell(index).smokeIndicator() < 100)
+                .forEach(index -> res.add(state.getCell(index)));
+        NeighbourUtils.neighboursStream(cellIndex, NeighbourUtils.Axis.Z)
+                .filter(state::hasCell)
+                .filter(index -> state.getCell(index).smokeIndicator() < 100)
+                .forEach(index -> res.add(state.getCell(index)));
+        return res;
+    }
+
+    private boolean isHorizontal(Vector3i cellIndex, Vector3i neighbourIndex) {
+        return NeighbourUtils.north(cellIndex).equals(neighbourIndex) ||
+                NeighbourUtils.south(cellIndex).equals(neighbourIndex) ||
+                NeighbourUtils.west(cellIndex).equals(neighbourIndex) ||
+                NeighbourUtils.east(cellIndex).equals(neighbourIndex);
+    }
+
+}
+
+class Int {
```

Comment:
I have no clue what is this class :P
You have to delete it, or if you sure we need that create better name for it and put it into separate file :P

Reply:
I changed the name, now it is more eloquent. No better idea to achive aim of usage.

Reply:
Yeah, I know this problem really well, it's kinda hard, because it makes a huge sense not to modify objects in lambdas, that's the entire idea of functional, non-side-effect programming :P
The case is we want to use it more like a imperative way.
Good enough solution is to finish your stream were you can, call `toList()` and after that iterate the list doing your side-effect actions. And use primitive `double` for that.